c++/src/capnp/compat/http-over-capnp.c++

```diff
@@ -578,6 +662,84 @@ public:
   }
 };
 
+class HttpOverCapnpFactory::HttpOverCapnpConnectResponseImpl : public kj::HttpService::ConnectResponse {
+public:
+  HttpOverCapnpConnectResponseImpl(
+      HttpOverCapnpFactory& factory, capnp::HttpService::ConnectParams::Reader request,
+      capnp::HttpService::ConnectClientRequestContext::Client context,
+      kj::Own<kj::AsyncInputStream> upStream, kj::Own<kj::AsyncIoStream> httpProxyStream) :
+      context(context), request(request), factory(factory), upStream(kj::mv(upStream)),
+      httpProxyStream(kj::mv(httpProxyStream)) {}
+
+  void accept(uint statusCode, kj::StringPtr statusText, const kj::HttpHeaders& headers) override {
+    KJ_REQUIRE(statusCode >= 200 && statusCode < 300, "the statusCode must be 2xx for accept");
+
+    auto req = context.startConnectRequest();
+    auto rpcResponse = req.initResponse();
+    rpcResponse.setStatusCode(statusCode);
+    rpcResponse.setStatusText(statusText);
+    rpcResponse.adoptHeaders(factory.headersToCapnp(
+        headers, Orphanage::getForMessageContaining(rpcResponse)));
+
+    kj::Own<kj::AsyncOutputStream> stream = factory.streamFactory.capnpToKj(request.getDown());
+    auto promises = kj::heapArrayBuilder<kj::Promise<void>>(3);
+    promises.add(req.send().ignoreResult());
+    // We write to the `down` pipe.
+    promises.add(httpProxyStream->pumpTo(*stream).then([req = kj::mv(request)](uint64_t) mutable {
+      auto endReq = req.getDown().endRequest();
+      return endReq.send().ignoreResult();
+    }).eagerlyEvaluate(nullptr));
```

Comment:
We should probably make sure that `down` gets dropped in the error case, which means you should attach `stream` before the `eagerlyEvaluate()` here, rather than after the `joinPromises()` later. That way it doesn't wait for the other direction to complete before dropping.

Separately, you should also call `httpProxyStream->abortRead()` on errors, otherwise the server app's attempts to write to the stream will just hang once the pump has stopped.

Reply:
Maybe this is the source of my issues